File: Tests/Main Tests/ProjectTestsVB/Test02.vb
Imports Tests.Core
Imports MSProject = NetOffice.MSProjectApi
Imports NetOffice.MSProjectApi.Enums

Public Class Test02
    Implements ITestPackage

    Public ReadOnly Property Description As String Implements Tests.Core.ITestPackage.Description
        Get
            Return "Test events."
        End Get
    End Property

    Public ReadOnly Property Language As String Implements Tests.Core.ITestPackage.Language
        Get
            Return "VB"
        End Get
    End Property

    Public ReadOnly Property Name As String Implements Tests.Core.ITestPackage.Name
        Get
            Return "Test02"
        End Get
    End Property

    Public ReadOnly Property OfficeProduct As String Implements Tests.Core.ITestPackage.OfficeProduct
        Get
            Return "Project"
        End Get
    End Property

    Private Sub ApplicationProjectTaskNewEvent(ByVal pj As MSProject.Project, ID As Integer)

        TaskNewEventCalled = True

    End Sub

    Private Sub ApplicationProjectBeforeTaskChangeEvent(ByVal tsk As MSProject.Task, ByVal Field As PjField, ByVal NewVal As Object, ByRef Cancel As Boolean)

        TaskChangeEventCalled = True

    End Sub

    Private Sub ApplicationProjectBeforeCloseEvent(ByVal pj As MSProject.Project, ByRef Cancel As Boolean)

        BeforeCloseEventCalled = True

    End Sub
     
    Private Sub ApplicationProjectBeforeTaskDeleteEvent(ByVal tsk As MSProject.Task, ByRef Cancel As Boolean)

        TaskDeleteEventCalled = True

    End Sub

  
    Public Function DoTest() As Tests.Core.TestResult Implements Tests.Core.ITestPackage.DoTest

        Dim application As MSProject.Application = Nothing
        Dim startTime As DateTime = DateTime.Now
        Try
            NetOffice.Settings.Default.MessageFilter.Enabled = True

            application = New MSProject.Application()

            Dim taskNewHandler As MSProject.Application_ProjectTaskNewEventHandler = AddressOf Me.ApplicationProjectTaskNewEvent
            AddHandler application.ProjectTaskNewEvent, taskNewHandler

            Dim projectBeforeCloseEventHandler As MSProject.Application_ProjectBeforeCloseEventHandler = AddressOf Me.ApplicationProjectBeforeCloseEvent
            AddHandler application.ProjectBeforeCloseEvent, projectBeforeCloseEventHandler

            Dim projectBeforeTaskChangeEventHandler As MSProject.Application_ProjectBeforeTaskChangeEventHandler = AddressOf Me.ApplicationProjectBeforeTaskChangeEvent
            AddHandler application.ProjectBeforeTaskChangeEvent, projectBeforeTaskChangeEventHandler

            Dim projectBeforeTaskDeleteHandler As MSProject.Application_ProjectBeforeTaskDeleteEventHandler = AddressOf Me.ApplicationProjectBeforeTaskDeleteEvent
            AddHandler application.ProjectBeforeTaskDeleteEvent, projectBeforeTaskDeleteHandler

            Dim newProject As MSProject.Project = application.Projects.Add()
            Dim task1 As MSProject.Task = newProject.Tasks.Add("Task 1")
            Dim task2 As MSProject.Task = newProject.Tasks.Add("Task 2")

            task2.Delete()
            application.FileCloseAll(False)

            RemoveHandler application.ProjectTaskNewEvent, taskNewHandler
            RemoveHandler application.ProjectBeforeCloseEvent, projectBeforeCloseEventHandler
            RemoveHandler application.ProjectBeforeTaskChangeEvent, projectBeforeTaskChangeEventHandler
            RemoveHandler application.ProjectBeforeTaskDeleteEvent, projectBeforeTaskDeleteHandler

            If (TaskDeleteEventCalled And TaskChangeEventCalled And BeforeCloseEventCalled And TaskNewEventCalled) Then
                Return New TestResult(True, DateTime.Now.Subtract(startTime), "", Nothing, "")
            Else
                Dim errorMessage As String = ""
                If (Not TaskDeleteEventCalled) Then errorMessage += "ProjectTaskNewEvent failed "
                If (Not TaskChangeEventCalled) Then errorMessage += "ProjectBeforeCloseEvent failed "
                If (Not BeforeCloseEventCalled) Then errorMessage += "ProjectBeforeTaskChangeEvent failed "
                If (Not TaskNewEventCalled) Then errorMessage += "ProjectBeforeTaskDeleteEvent failed "
                Return New TestResult(False, DateTime.Now.Subtract(startTime), errorMessage, Nothing, "")
            End If

        Catch ex As Exception

            Return New TestResult(False, DateTime.Now.Subtract(startTime), ex.Message, ex, "")

        Finally

            NetOffice.Settings.Default.MessageFilter.Enabled = False

            If Not IsNothing(application) Then
                application.Quit(False)
                application.Dispose()
            End If

        End Try

    End Function

    Dim TaskDeleteEventCalled As Boolean
    Dim TaskChangeEventCalled As Boolean
    Dim BeforeCloseEventCalled As Boolean
    Dim TaskNewEventCalled As Boolean

End Class

File: Tests/Main Tests/ProjectTestsCSharp/Test02.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Tests.Core;
using MSProject = NetOffice.MSProjectApi;
using NetOffice.MSProjectApi.Enums;

namespace ProjectTestsCSharp
{
    public class Test02 : ITestPackage
    {
        #region TestPackage Member

        public string Name
        {
            get { return "Test02"; }
        }

        public string Description
        {
            get { return "Test events"; }
        }

        public string OfficeProduct
        {
            get { return "Project"; }
        }

        public string Language
        {
            get { return "C#"; }
        }

        public TestResult DoTest()
        {
            NetOffice.Settings.Default.MessageFilter.Enabled = true;
            MSProject.Application application = null;
            DateTime startTime = DateTime.Now;
            try
            {
                application = new MSProject.Application();

                application.ProjectTaskNewEvent += new MSProject.Application_ProjectTaskNewEventHandler(ApplicationProjectTaskNewEvent);
                application.ProjectBeforeCloseEvent += new MSProject.Application_ProjectBeforeCloseEventHandler(ApplicationProjectBeforeCloseEvent);
                application.ProjectBeforeTaskChangeEvent += new MSProject.Application_ProjectBeforeTaskChangeEventHandler(ApplicationProjectBeforeTaskChangeEvent);
                application.ProjectBeforeTaskDeleteEvent += new MSProject.Application_ProjectBeforeTaskDeleteEventHandler(ApplicationProjectBeforeTaskDeleteEvent);
               
                MSProject.Project newProject = application.Projects.Add();
                MSProject.Task task1 = newProject.Tasks.Add("Task 1");
                MSProject.Task task2 =  newProject.Tasks.Add("Task 2");
                
                task2.Delete();

                application.FileCloseAll(PjSaveType.pjDoNotSave);

                if (TaskDeleteEventCalled && TaskChangeEventCalled && BeforeCloseEventCalled && TaskChangeEventCalled)
                    return new TestResult(true, DateTime.Now.Subtract(startTime), "", null, "");
                else 
                {
                    string errorMessage = "";
                    if (!TaskDeleteEventCalled)
                        errorMessage += "ProjectBeforeTaskDeleteEvent failed ";
                    if (!TaskChangeEventCalled)
                        errorMessage += "ProjectBeforeTaskChangeEvent failed ";
                    if (!BeforeCloseEventCalled)
                        errorMessage += "ProjectBeforeCloseEvent failed ";
                    if (!TaskNewEventCalled)
                        errorMessage += "ProjectTaskNewEvent failed ";

                    return new TestResult(false, DateTime.Now.Subtract(startTime), errorMessage, null, "");
                }
            }
            catch (Exception exception)
            {
                return new TestResult(false, DateTime.Now.Subtract(startTime), exception.Message, exception, "");
            }
            finally
            {
                NetOffice.Settings.Default.MessageFilter.Enabled = false;

                if (null != application)
                {
                    application.Quit(PjSaveType.pjDoNotSave);
                    application.Dispose();
                }
            }
        }

        #endregion

        private bool TaskDeleteEventCalled { get; set; }
        private bool TaskChangeEventCalled { get; set; }
        private bool BeforeCloseEventCalled { get; set; }
        private bool TaskNewEventCalled { get; set; }


        void ApplicationProjectBeforeTaskDeleteEvent(MSProject.Task tsk, ref bool Cancel)
        {
            TaskDeleteEventCalled = true;
        }

        void ApplicationProjectBeforeTaskChangeEvent(MSProject.Task tsk, PjField Field, object NewVal, ref bool Cancel)
        {
            TaskChangeEventCalled = true;
        }

        void ApplicationProjectBeforeCloseEvent(MSProject.Project pj, ref bool Cancel)
        {
            BeforeCloseEventCalled = true;
        }

        void ApplicationProjectTaskNewEvent(MSProject.Project pj, int ID)
        {
            TaskNewEventCalled = true;
        }
    }
}
